File: backend/app/schemas/reminders.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class ResponseModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class RequestModel(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)


class ReminderListQuery(RequestModel):
    status: str | None = None
    source_type: str | None = None
    item_id: int | None = None
    loan_id: int | None = None
    overdue: bool | None = None
    upcoming_days: int | None = Field(default=None, ge=1, le=90)
    include_archived: bool = False
    search: str | None = None
    sort: str = "due_asc"
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


class ReminderCreate(RequestModel):
    title: str = Field(min_length=1, max_length=160)
    description: str = ""
    item_id: int | None = None
    due_date: date | None = None
    remind_at: date | None = None
    priority: str = "normal"


class ReminderUpdate(RequestModel):
    title: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = None
    item_id: int | None = None
    due_date: date | None = None
    remind_at: date | None = None
    priority: str | None = None


class ReminderItemSummary(ResponseModel):
    id: int
    name: str
    is_archived: bool


class ReminderLoanSummary(ResponseModel):
    id: int
    borrower_name: str
    expected_return_date: date | None = None
    returned_at: datetime | None = None


class ReminderSummaryResponse(ResponseModel):
    id: int
    title: str
    description: str
    source_type: str
    item_id: int | None = None
    loan_id: int | None = None
    due_date: date | None = None
    remind_at: date | None = None
    status: str
    priority: str
    due_state: str
    days_until_due: int | None = None
    created_by_user_id: int | None = None
    completed_by_user_id: int | None = None
    dismissed_by_user_id: int | None = None
    completed_at: datetime | None = None
    dismissed_at: datetime | None = None
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    item: ReminderItemSummary | None = None
    loan: ReminderLoanSummary | None = None


class ReminderDetailResponse(ReminderSummaryResponse):
    pass


class ReminderListResponse(ResponseModel):
    items: list[ReminderSummaryResponse] = Field(default_factory=list)
    total: int
    page: int
    page_size: int

File: frontend/src/api/reminders.ts
import { apiClient } from './client'

export type ReminderStatus = 'pending' | 'done' | 'dismissed'
export type ReminderSourceType = 'manual' | 'loan_return'
export type ReminderPriority = 'low' | 'normal' | 'high'
export type ReminderDueState = ReminderStatus | 'overdue' | 'upcoming' | 'scheduled' | 'none'
export type ReminderSort =
  | 'due_asc'
  | 'due_desc'
  | 'created_asc'
  | 'created_desc'
  | 'updated_asc'
  | 'updated_desc'
  | 'priority_desc'

export interface ReminderItemSummary {
  id: number
  name: string
  is_archived: boolean
}

export interface ReminderLoanSummary {
  id: number
  borrower_name: string
  expected_return_date: string | null
  returned_at: string | null
}

export interface ReminderSummary {
  id: number
  title: string
  description: string
  source_type: ReminderSourceType
  item_id: number | null
  loan_id: number | null
  due_date: string | null
  remind_at: string | null
  status: ReminderStatus
  priority: ReminderPriority
  due_state: ReminderDueState
  days_until_due: number | null
  created_by_user_id: number | null
  completed_by_user_id: number | null
  dismissed_by_user_id: number | null
  completed_at: string | null
  dismissed_at: string | null
  archived_at: string | null
  created_at: string
  updated_at: string
  item: ReminderItemSummary | null
  loan: ReminderLoanSummary | null
}

export interface ReminderDetail extends ReminderSummary {}

export interface ReminderListResponse {
  items: ReminderSummary[]
  total: number
  page: number
  page_size: number
}

export interface ReminderFilters {
  status?: ReminderStatus | null
  source_type?: ReminderSourceType | null
  item_id?: number | null
  loan_id?: number | null
  overdue?: boolean | null
  upcoming_days?: number | null
  include_archived?: boolean
  search?: string | null
  sort?: ReminderSort
  page?: number
  page_size?: number
}

export interface ReminderCreateRequest {
  title: string
  description?: string
  item_id?: number | null
  due_date?: string | null
  remind_at?: string | null
  priority?: ReminderPriority
}

export interface ReminderUpdateRequest {
  title?: string | null
  description?: string | null
  item_id?: number | null
  due_date?: string | null
  remind_at?: string | null
  priority?: ReminderPriority | null
}

export async function listRemindersApi(filters: ReminderFilters = {}): Promise<ReminderListResponse> {
  const response = await apiClient.get<ReminderListResponse>('/reminders', { params: filters })
  return response.data
}

export async function createReminderApi(payload: ReminderCreateRequest): Promise<ReminderDetail> {
  const response = await apiClient.post<ReminderDetail>('/reminders', payload)
  return response.data
}

export async function fetchReminderDetailApi(reminderId: number): Promise<ReminderDetail> {
  const response = await apiClient.get<ReminderDetail>(`/reminders/${reminderId}`)
  return response.data
}

export async function updateReminderApi(
  reminderId: number,
  payload: ReminderUpdateRequest
): Promise<ReminderDetail> {
  const response = await apiClient.patch<ReminderDetail>(`/reminders/${reminderId}`, payload)
  return response.data
}

export async function completeReminderApi(reminderId: number): Promise<ReminderDetail> {
  const response = await apiClient.post<ReminderDetail>(`/reminders/${reminderId}/complete`)
  return response.data
}

export async function dismissReminderApi(reminderId: number): Promise<ReminderDetail> {
  const response = await apiClient.post<ReminderDetail>(`/reminders/${reminderId}/dismiss`)
  return response.data
}

export async function reopenReminderApi(reminderId: number): Promise<ReminderDetail> {
  const response = await apiClient.post<ReminderDetail>(`/reminders/${reminderId}/reopen`)
  return response.data
}

export async function archiveReminderApi(reminderId: number): Promise<ReminderDetail> {
  const response = await apiClient.delete<ReminderDetail>(`/reminders/${reminderId}`)
  return response.data
}
